subsys/testsuite/ztest/include/ztest_test.h

```diff
@@ -27,7 +29,65 @@ struct unit_test {
 	uint32_t thread_options;
 };
 
-void z_ztest_run_test_suite(const char *name, struct unit_test *suite);
+/**
+ * Metadata about a ztest suite
+ */
+struct ztest_suite_metadata {
+	uint32_t run_count;
+	uint32_t skip_count;
+	uint32_t fail_count;
+};
+
+/**
+ * A single node of test suite. Each node should be added to unit_test_head to
+ * form a linked list that can be run via ztest_run_registered_test_suites().
+ */
+struct ztest_suite_node {
+	/** The name of the test suite. */
+	const char *name;
+	/** Pointer to the test suite. */
+	struct unit_test *suite;
+	/**
+	 * An optional pragma function to determine if the test should run. If NULL, then the test
+	 * will only run once on the first attempt.
+	 */
+	bool (*pragma)(const void *state);
+	/** Metadata */
+	struct ztest_suite_metadata metadata;
+};
+
+extern const struct ztest_suite_node __ztest_suite_start[];
+extern const struct ztest_suite_node __ztest_suite_end[];
+
+/**
+ * Create and register a ztest suite. Using this macro creates a new test suite (using
+ * ztest_test_suite). It then creates a struct unit_test_node and an initializer function which will
+ * be run via a SYS_INIT call at the APPLICATION initialization level and priority 99.
+ *
+ * Tests can then be run by calling ztest_run_registered_test_suites(const void *state) by passing
+ * in the current state. See the documentation for ztest_run_registered_test_suites for more info.
+ *
+ * @param SUITE_NAME The name of the suite (see ztest_test_suite for more info)
+ * @param PRAGMA A function to test against the state and determine if the test should run.
+ */
+#define ztest_register_test_suite(SUITE_NAME, PRAGMA, ...)                                         \
+	ztest_test_suite(SUITE_NAME, __VA_ARGS__);                                                 \
+	static struct ztest_suite_node z_ztest_test_node_##SUITE_NAME                              \
+		__attribute__((__section__(".ztest_suites"))) __used = {                           \
```

Comment:
Should use an abstraction like `Z_GENERIC_SECTION` rather than gcc-specific attributes.